pkg/master/controller/crdregistration/crdregistration_controller.go

```diff
@@ -195,26 +198,25 @@ func (c *crdRegistrationController) handleVersionUpdate(groupVersion schema.Grou
 		return err
 	}
 	for _, crd := range crds {
-		if crd.Spec.Version == groupVersion.Version && crd.Spec.Group == groupVersion.Group {
-			found = true
-			break
+		if crd.Spec.Group != groupVersion.Group {
+			continue
+		}
+		for _, version := range crd.Spec.Versions {
+			if version.Name == groupVersion.Version && version.Served {
+				c.apiServiceRegistration.AddAPIServiceToSync(&apiregistration.APIService{
+					ObjectMeta: metav1.ObjectMeta{Name: apiServiceName},
+					Spec: apiregistration.APIServiceSpec{
+						Group:                groupVersion.Group,
+						Version:              groupVersion.Version,
+						GroupPriorityMinimum: 1000, // CRDs should have relatively low priority
+						VersionPriority:      100,  // CRDs should have relatively low priority
+					},
+				})
+				return nil
+			}
 		}
 	}
 
-	if !found {
-		c.apiServiceRegistration.RemoveAPIServiceToSync(apiServiceName)
-		return nil
-	}
-
-	c.apiServiceRegistration.AddAPIServiceToSync(&apiregistration.APIService{
-		ObjectMeta: metav1.ObjectMeta{Name: apiServiceName},
-		Spec: apiregistration.APIServiceSpec{
-			Group:                groupVersion.Group,
-			Version:              groupVersion.Version,
-			GroupPriorityMinimum: 1000, // CRDs should have relatively low priority
-			VersionPriority:      100,  // CRDs should have relatively low priority
-		},
-	})
-
+	c.apiServiceRegistration.RemoveAPIServiceToSync(apiServiceName)
```

Comment:
does this remove properly if there are multiple versions?

Reply:
It may not but the customresource_handler.ServeHTTP only serve it if the version exists and delegate the rest. I don't like it too, will look into an alternative. While it should not functionally change anything, I would like to, at least, remove API services for disabled/removed versions.

Reply:
> While it should not functionally change anything, I would like to, at least, remove API services for disabled/removed versions.

APIService objects drive discovery docs, so we need them to be correct. make sure to include checks of discovery docs in the version lifecycle test described below

Reply:
Fixed.